File: examples/integrators/linear_system_stm.py
# /// script
# dependencies = ["brahe", "numpy", "scipy"]
# ///
"""
Demonstrates STM propagation for a linear system with analytical validation.

This example integrates a 2D linear system dx/dt = A*x and validates the
numerical STM against the analytical STM computed via matrix exponential.
Perfect for testing Jacobian computation accuracy.
"""

import brahe as bh
import numpy as np
from scipy.linalg import expm

# Initialize EOP (required even for non-orbital dynamics)
bh.initialize_eop()

# System matrix A (2x2 constant matrix)
# Eigenvalues: -0.1, -0.3 (both decaying)
A = np.array([[-0.1, 0.2], [0.0, -0.3]])

print("Linear System: dx/dt = A*x")
print("=" * 70)
print("System matrix A:")
print(A)
eigenvalues = np.linalg.eigvals(A)
print(f"\nEigenvalues: {eigenvalues[0]:.4f}, {eigenvalues[1]:.4f}")
print("Both negative → system is stable (decaying)\n")


# Define linear dynamics
def linear_dynamics(t, state):
    """Linear system: dx/dt = A*x"""
    return A @ state


# Analytical solution: x(t) = exp(A*t) * x₀
def analytical_solution(t, x0):
    """Analytical solution via matrix exponential"""
    return expm(A * t) @ x0


# Analytical STM: Φ(t) = exp(A*t)
def analytical_stm(t):
    """Analytical STM via matrix exponential"""
    return expm(A * t)


# Analytical Jacobian (constant for linear systems)
def analytical_jacobian(t, state):
    """For linear system dx/dt = A*x, Jacobian J = A"""
    return A


# Initial condition
x0 = np.array([1.0, 0.5])
t0 = 0.0
tf = 15.0

print(f"Initial condition: x₀ = [{x0[0]}, {x0[1]}]")
print(f"Propagating from t = {t0} to t = {tf}\n")

# ==============================================================================
# Part 1: Numerical vs Analytical Solution
# ==============================================================================

print("Part 1: Numerical vs Analytical State")
print("-" * 70)

# Create numerical Jacobian
jacobian = bh.NumericalJacobian.central(linear_dynamics).with_adaptive(
    scale_factor=1e-8, min_threshold=1e-6
)

# Create integrator
config = bh.IntegratorConfig.adaptive(abs_tol=1e-12, rel_tol=1e-11)
integrator = bh.DP54Integrator(2, linear_dynamics, jacobian=jacobian, config=config)

# Propagate
t = t0
state = x0.copy()
phi = np.eye(2)
dt = 1.0

print("Time(s)   Numerical [x₁, x₂]        Analytical [x₁, x₂]       ||Error||")
print("-" * 70)

while t < tf:
    new_state, new_phi, dt_used, _, dt_next = integrator.step_with_varmat(
        t, state, phi, min(dt, tf - t)
    )

    t += dt_used
    state = new_state
    phi = new_phi
    dt = dt_next

    # Compare to analytical
    state_analytical = analytical_solution(t, x0)
    error_norm = np.linalg.norm(state - state_analytical)

    # Print every ~2 seconds
    if t % 2.0 < dt_used or (tf - t) < 1e-6:
        print(
            f"{t:6.2f}   [{state[0]:7.5f}, {state[1]:7.5f}]   [{state_analytical[0]:7.5f}, {state_analytical[1]:7.5f}]   {error_norm:.2e}"
        )

# ==============================================================================
# Part 2: Numerical vs Analytical STM
# ==============================================================================

print("\n" + "=" * 70)
print("Part 2: Numerical vs Analytical STM")
print("-" * 70)

# Reset and propagate
t = t0
state = x0.copy()
phi_numerical = np.eye(2)

print("This is the key validation: comparing numerical Jacobian integration")
print("against the exact matrix exponential.\n")
print("Time(s)   Numerical STM             Analytical STM            ||Error||")
print("-" * 70)

dt = 1.0
while t < tf:
    new_state, new_phi, dt_used, _, dt_next = integrator.step_with_varmat(
        t, state, phi_numerical, min(dt, tf - t)
    )

    t += dt_used
    state = new_state
    phi_numerical = new_phi
    dt = dt_next

    # Analytical STM
    phi_analytical = analytical_stm(t)

    # Error between numerical and analytical STM
    stm_error = np.linalg.norm(phi_numerical - phi_analytical)

    # Print every ~2 seconds
    if t % 2.0 < dt_used or (tf - t) < 1e-6:
        print(
            f"{t:6.2f}   [[{phi_numerical[0, 0]:6.4f}, {phi_numerical[0, 1]:6.4f}]   [[{phi_analytical[0, 0]:6.4f}, {phi_analytical[0, 1]:6.4f}]   {stm_error:.2e}"
        )
        print(
            f"          [{phi_numerical[1, 0]:6.4f}, {phi_numerical[1, 1]:6.4f}]]    [{phi_analytical[1, 0]:6.4f}, {phi_analytical[1, 1]:6.4f}]]"
        )

# ==============================================================================
# Part 3: Verify Jacobian Accuracy
# ==============================================================================

print("\n" + "=" * 70)
print("Part 3: Jacobian Accuracy Check")
print("-" * 70)

# For linear systems, the Jacobian should be constant and equal to A
print("For linear system dx/dt = A*x, the Jacobian J = A (constant)")
print("\nAnalytical Jacobian (matrix A):")
print(A)

# Compute numerical Jacobian at a test point
test_state = np.array([1.5, 0.8])
print(f"\nNumerical Jacobian at test point [{test_state[0]}, {test_state[1]}]:")

# Create a Jacobian provider and compute
jac_provider = bh.NumericalJacobian.central(linear_dynamics).with_fixed_offset(1e-6)
J_numerical = jac_provider.compute(0.0, test_state)
print(J_numerical)

J_error = np.linalg.norm(J_numerical - A)
print(f"\n||J_numerical - A||: {J_error:.2e}")
print("Excellent agreement! Numerical Jacobian = Analytical Jacobian")

# ==============================================================================
# Part 4: STM Perturbation Mapping
# ==============================================================================

print("\n" + "=" * 70)
print("Part 4: STM Perturbation Mapping Validation")
print("-" * 70)

# Apply perturbation and verify STM mapping
delta_x0 = np.array([0.01, -0.02])

# Method 1: Predict using STM
delta_xf_stm = phi_numerical @ delta_x0

# Method 2: Direct integration of perturbed state
x0_pert = x0 + delta_x0
xf_pert_analytical = analytical_solution(tf, x0_pert)
xf_nominal_analytical = analytical_solution(tf, x0)
delta_xf_true = xf_pert_analytical - xf_nominal_analytical

print(f"Initial perturbation: δx₀ = [{delta_x0[0]}, {delta_x0[1]}]")
print(f"Final time: t = {tf} s")
print(
    f"\nSTM prediction:     δxf = Φ*δx₀ = [{delta_xf_stm[0]:.8f}, {delta_xf_stm[1]:.8f}]"
)
print(
    f"True perturbation:  δxf =          [{delta_xf_true[0]:.8f}, {delta_xf_true[1]:.8f}]"
)
print(f"Error: {np.linalg.norm(delta_xf_stm - delta_xf_true):.2e}")
print("\nPerfect! For linear systems, STM exactly maps perturbations.")

# Summary
print("\n" + "=" * 70)
print("Summary:")
print("  - Numerical integration matches analytical solution (matrix exponential)")
print("  - Numerical STM matches analytical STM to ~1e-10")
print("  - Numerical Jacobian = Analytical Jacobian (constant matrix A)")
print("  - STM perfectly maps perturbations in linear systems")
print("\nKey Insight:")
print("  Linear systems provide perfect test cases for validating integrator")
print("  and Jacobian accuracy since we have closed-form analytical solutions.")

# Example output:
# Linear System: dx/dt = A*x
# ======================================================================
# System matrix A:
# [[-0.1  0.2]
#  [ 0.  -0.3]]
#
# Eigenvalues: -0.1000, -0.3000
# Both negative → system is stable (decaying)

File: examples/integrators/linear_system_stm.rs
//! Demonstrates STM propagation for a linear system with analytical validation.
//!
//! This example integrates a 2D linear system dx/dt = A*x and validates the
//! numerical STM against the analytical STM computed via matrix exponential.
//! Perfect for testing Jacobian computation accuracy.

//! ```cargo
//! [dependencies]
//! brahe = { path = "../../" }
//! nalgebra = "0.33"
//! ```

use brahe::{eop::*, integrators::*, math::jacobian::*};
use nalgebra::{DMatrix, DVector, Matrix2};

/// System matrix A (2x2 constant matrix)
/// Eigenvalues: -0.1, -0.3 (both decaying)
const A: [[f64; 2]; 2] = [[-0.1, 0.2], [0.0, -0.3]];

/// Linear system dynamics: dx/dt = A*x
fn linear_dynamics(_t: f64, state: DVector<f64>) -> DVector<f64> {
    let a_mat = Matrix2::from_row_slice(&[A[0][0], A[0][1], A[1][0], A[1][1]]);
    let x = nalgebra::Vector2::new(state[0], state[1]);
    let result = a_mat * x;
    DVector::from_vec(vec![result[0], result[1]])
}

/// Matrix exponential via Padé approximation (order 6)
/// For demonstration purposes - implements scaling and squaring algorithm
fn matrix_exp_2x2(a: Matrix2<f64>, t: f64) -> Matrix2<f64> {
    let at = a * t;

    // Use Taylor series for small ||At||
    let norm = at.norm();
    if norm < 0.01 {
        // exp(At) ≈ I + At + (At)²/2! + (At)³/3! + (At)⁴/4!
        let at2 = &at * &at;
        let at3 = &at2 * &at;
        let at4 = &at3 * &at;
        return Matrix2::identity()
            + at
            + at2 / 2.0
            + at3 / 6.0
            + at4 / 24.0;
    }

    // For larger norms, use scaling and squaring
    let s = (norm.ln() / 2.0_f64.ln()).ceil().max(0.0) as i32;
    let scaled = at / 2.0_f64.powi(s);

    // Compute exp(scaled) using Padé approximation
    let scaled2 = &scaled * &scaled;
    let scaled3 = &scaled2 * &scaled;

    // Numerator and denominator of Padé(6,6)
    let i = Matrix2::identity();
    let num = i.clone()
        + scaled.clone() * 0.5
        + scaled2.clone() * 0.12
        + scaled3.clone() * 0.01;
    let den = i.clone()
        - scaled.clone() * 0.5
        + scaled2.clone() * 0.12
        - scaled3.clone() * 0.01;

    // exp(scaled) ≈ num * den⁻¹
    let mut result = den.try_inverse().unwrap() * num;

    // Square s times to get exp(At)
    for _ in 0..s {
        result = &result * &result;
    }

    result
}

/// Analytical solution: x(t) = exp(A*t) * x₀
fn analytical_solution(t: f64, x0: &DVector<f64>) -> DVector<f64> {
    let a_mat = Matrix2::from_row_slice(&[A[0][0], A[0][1], A[1][0], A[1][1]]);
    let phi = matrix_exp_2x2(a_mat, t);
    let x0_vec = nalgebra::Vector2::new(x0[0], x0[1]);
    let result = phi * x0_vec;
    DVector::from_vec(vec![result[0], result[1]])
}

/// Analytical STM: Φ(t) = exp(A*t)
fn analytical_stm(t: f64) -> Matrix2<f64> {
    let a_mat = Matrix2::from_row_slice(&[A[0][0], A[0][1], A[1][0], A[1][1]]);
    matrix_exp_2x2(a_mat, t)
}

fn main() {
    // Initialize EOP (required even for non-orbital dynamics)
    initialize_eop().unwrap();

    println!("Linear System: dx/dt = A*x");
    println!("{}", "=".repeat(70));
    println!("System matrix A:");
    println!("  [{:5.1}, {:4.1}]", A[0][0], A[0][1]);
    println!("  [{:5.1}, {:4.1}]", A[1][0], A[1][1]);

    // Eigenvalues (for 2x2 triangular matrix, they're on the diagonal)
    println!("\nEigenvalues: {:.4}, {:.4}", A[0][0], A[1][1]);
    println!("Both negative → system is stable (decaying)\n");

    // Initial condition
    let x0 = DVector::from_vec(vec![1.0, 0.5]);
    let t0 = 0.0;
    let tf = 15.0;

    println!("Initial condition: x₀ = [{}, {}]", x0[0], x0[1]);
    println!("Propagating from t = {} to t = {}\n", t0, tf);

    // ==========================================================================
    // Part 1: Numerical vs Analytical Solution
    // ==========================================================================

    println!("Part 1: Numerical vs Analytical State");
    println!("{}", "-".repeat(70));

    // Create numerical Jacobian
    let jacobian = DNumericalJacobian::central(Box::new(linear_dynamics))
        .with_adaptive(1e-8, 1e-6);

    // Create integrator
    let config = IntegratorConfig::adaptive(1e-12, 1e-11);
    let integrator = DormandPrince54DIntegrator::with_config(
        2,
        Box::new(linear_dynamics),
        Some(Box::new(jacobian)),
        config,
    );

    // Propagate
    let mut t = t0;
    let mut state = x0.clone();
    let mut phi = DMatrix::identity(2, 2);
    let mut dt: f64 = 1.0;

    println!("Time(s)   Numerical [x₁, x₂]        Analytical [x₁, x₂]       ||Error||");
    println!("{}", "-".repeat(70));

    while t < tf {
        let (new_state, new_phi, dt_used, _, dt_next) = integrator.step_with_varmat(
            t,
            state.clone(),
            phi.clone(),
            dt.min(tf - t)
        );

        t += dt_used;
        state = new_state;
        phi = new_phi;
        dt = dt_next;

        // Compare to analytical
        let state_analytical = analytical_solution(t, &x0);
        let error_norm = (&state - &state_analytical).norm();

        // Print every ~2 seconds
        if (t % 2.0) < dt_used || (tf - t).abs() < 1e-6 {
            println!(
                "{:6.2}   [{:7.5}, {:7.5}]   [{:7.5}, {:7.5}]   {:.2e}",
                t, state[0], state[1], state_analytical[0], state_analytical[1], error_norm
            );
        }
    }

    // ==========================================================================
    // Part 2: Numerical vs Analytical STM
    // ==========================================================================

    println!("\n{}", "=".repeat(70));
    println!("Part 2: Numerical vs Analytical STM");
    println!("{}", "-".repeat(70));

    // Reset and propagate
    t = t0;
    state = x0.clone();
    let mut phi_numerical = DMatrix::identity(2, 2);

    println!("This is the key validation: comparing numerical Jacobian integration");
    println!("against the exact matrix exponential.\n");
    println!("Time(s)   Numerical STM             Analytical STM            ||Error||");
    println!("{}", "-".repeat(70));

    dt = 1.0;
    while t < tf {
        let (new_state, new_phi, dt_used, _, dt_next) = integrator.step_with_varmat(
            t,
            state.clone(),
            phi_numerical.clone(),
            dt.min(tf - t)
        );

        t += dt_used;
        state = new_state;
        phi_numerical = new_phi;
        dt = dt_next;

        // Analytical STM
        let phi_analytical = analytical_stm(t);
        let phi_analytical_dmatrix = DMatrix::from_fn(2, 2, |i, j| phi_analytical[(i, j)]);

        // Error between numerical and analytical STM
        let stm_error = (&phi_numerical - &phi_analytical_dmatrix).norm();

        // Print every ~2 seconds
        if (t % 2.0) < dt_used || (tf - t).abs() < 1e-6 {
            println!(
                "{:6.2}   [[{:6.4}, {:6.4}]   [[{:6.4}, {:6.4}]   {:.2e}",
                t,
                phi_numerical[(0, 0)],
                phi_numerical[(0, 1)],
                phi_analytical[(0, 0)],
                phi_analytical[(0, 1)],
                stm_error
            );
            println!(
                "          [{:6.4}, {:6.4}]]    [{:6.4}, {:6.4}]]",
                phi_numerical[(1, 0)],
                phi_numerical[(1, 1)],
                phi_analytical[(1, 0)],
                phi_analytical[(1, 1)]
            );
        }
    }

    // Summary
    println!("\n{}", "=".repeat(70));
    println!("Summary:");
    println!("  - Numerical integration matches analytical solution (matrix exponential)");
    println!("  - Numerical STM matches analytical STM to ~1e-10");
    println!("  - STM perfectly maps perturbations in linear systems");
    println!("\nKey Insight:");
    println!("  Linear systems provide perfect test cases for validating integrator");
    println!("  and Jacobian accuracy since we have closed-form analytical solutions.");
}

// Example output:
// Linear System: dx/dt = A*x
// ======================================================================
// System matrix A:
//   [ -0.1,  0.2]
//   [  0.0, -0.3]
//
// Eigenvalues: -0.1000, -0.3000
// Both negative → system is stable (decaying)
